MobileWallet/Screens/Send/AddRecipient/AddRecipientViewController.swift

```diff
@@ -40,45 +40,133 @@
 
 import UIKit
 
-class AddRecipientViewController: UIViewController, ContactsTableDelegate {
+class AddRecipientViewController: UIViewController, ContactsTableDelegate, UITextFieldDelegate, ScanViewControllerDelegate {
     private let SIDE_PADDING = Theme.shared.sizes.appSidePadding
     private let INPUT_CORNER_RADIUS: CGFloat = 6
     private let INPUT_CONTAINER_HEIGHT: CGFloat = 90
 
     private let inputContainerView = UIView()
     private let inputBox = UITextField()
     private let scanButton = QRButton()
+    private let continueButton = ActionButton()
+    private var continueButtonBottomConstraint = NSLayoutConstraint()
+    private let contactsTableVC = ContactsTableViewController(style: .grouped)
+    private let pasteEmojisView = PasteEmojisView()
+    private var pasteEmojisViewBottomAnchorConstraint = NSLayoutConstraint()
+    private let dimView = UIView()
 
-    private let contactsTableVC = ContactsTableViewController()
+    private var isEditingSearchBox: Bool = false {
+        didSet {
+            if isEditingSearchBox {
+                inputBox.becomeFirstResponder()
+            } else {
+                inputBox.resignFirstResponder()
+            }
+        }
+    }
+
+    private var isShowingContinueButton: Bool = false {
+        didSet {
+            if isShowingContinueButton {
+                contactsTableVC.tableView.removeFromSuperview()
+
+                continueButtonBottomConstraint.isActive = false
+
+                UIView.animate(withDuration: 0.5) {
+                    self.continueButtonBottomConstraint = self.continueButton.bottomAnchor.constraint(equalTo: self.view.bottomAnchor, constant: -self.SIDE_PADDING)
+                    self.continueButtonBottomConstraint.isActive = true
+                    self.inputContainerView.layer.shadowOpacity = 0.1
+                    self.view.layoutIfNeeded()
+                }
+            } else {
+                continueButtonBottomConstraint.isActive = false
+                continueButtonBottomConstraint = continueButton.bottomAnchor.constraint(equalTo: view.bottomAnchor, constant: 100)
+                continueButtonBottomConstraint.isActive = true
+            }
+        }
+    }
+
+    private var clipboardEmojis: String = "" {
+        didSet {
+            if clipboardEmojis.isEmpty {
+                pasteEmojisView.isHidden = true
+            } else {
+                pasteEmojisView.setEmojis(emojis: clipboardEmojis) {
```

Comment:
I haven't tested this but this looks like it would cause a retain cycle, the closure is a property on pasteEmojisView and here it references addrecipientvc

Reply:
Can't seem to do a retain cycle check here. Compiler error I get `Initializer for conditional binding must have Optional type`.

This closure only gets called when the user taps the emoji view, once the VC has been deconstructed they wouldn't be able to tap it. Would this still cause a memory leak?

Reply:
Like I said I haven't tested it with Instruments so I'm not entirely sure but I think it still would because they are strongly referencing each other. You can test this with instruments or you can check if deinit is called on the vc by implementing the deinit function and checking if it's called when you pop out of the vc.

Reply:
@semihcihan thanks I didn't know about the memory graph tool. That helped. All been patched.